### DISCUSSION TITLE
Adding a specific EMF-ECORE and a customized palette

### DISCUSSION THREAD

**Discussion by Mohammad-Si** [eclipse-glsp/glsp#1311] [Q&A]
Hi Guys,

I have successfully cloned both the [glsp-server (Java)](https://github.com/eclipse-glsp/glsp-server) and [glsp-theia-integration](https://github.com/eclipse-glsp/glsp-theia-integration) for both server and client sides subsequently. epositories for both server and client sides, respectively. Currently, I am seeking guidance on integrating an additional ECORE model into the glsp-server framework as a new module.
Despite reviewing the guidelines like [this](https://eclipse.dev/glsp/documentation/sourcemodel/) and engaging in community discussions, I am uncertain about the specific procedures required to achieve this goal in a clear and systematic manner. My ultimate objective is to visualize the changes on localhost:3000 with a customized palette.
What should I do exactly on the server-side?
Do I need any change on the client-side?
I would greatly appreciate any assistance or guidance you can provide in navigating this process. 
Thank you for your support.

**Comment by planger** [comment]
Hi @Mohammad-Si,

Great to hear that you’re working on a GLSP diagram editor for an EMF model!

Can you please elaborate on your goal? Do you need to load an additional EMF model for an existing diagram editor -- ie as an additional data source for your diagram editor? Or do you want to create a separate diagram editor that is fueled by another EMF model?
If it is a separate diagram editor, do those two EMF models depend on each other, or can they live separately?
Answering those questions would help me to give more concrete suggestions.

Is there a reason why you cloned the GLSP server framework repository? Usually, we recommend to start from the GLSP project template, in your case, this would be the one for Theia + Java-EMF server: https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-theia
Do load additional or other EMF models (conforming to different Ecores), you'd have to adapt the source model storage:
https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/model/TaskListSourceModelStorage.java

To translate this model now into a GModel, you'd have to adapt the GModelFactory:
https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/model/TaskListGModelFactory.java

To handle edit operations and apply them to your source model, you'd have to provide the respective handlers:
https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/handler

These handlers would have to be registered in the diagram module:
https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/TaskListDiagramModule.java#L67

Maybe helpful as well, if you are up to handle multiple diagram editors in one server: https://github.com/eclipse-glsp/glsp/discussions/1285

Best wishes,

Philip
